dial/throttler.go

```diff
@@ -0,0 +1,178 @@
+package dial
+
+import (
+	"context"
+	"sync"
+
+	addrutil "github.com/libp2p/go-addr-util"
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+// ConcurrentFdDials is the number of concurrent outbound dials over transports
+// that consume file descriptors
+const ConcurrentFdDials = 160
+
+// DefaultPerPeerRateLimit is the number of concurrent outbound dials to make
+// per peer
+const DefaultPerPeerRateLimit = 8
+
+type throttler struct {
+	lk sync.Mutex
+
+	fdConsuming int
+	fdLimit     int
+	waitingOnFd []*Job
+
+	activePerPeer      map[peer.ID]int
+	perPeerLimit       int
+	waitingOnPeerLimit map[peer.ID][]*Job
+
+	finishedCh chan *Job
+
+	peerTokenFreed chan peer.ID
+	fdTokenFreed   chan struct{}
+
+	inCh         <-chan *Job
+	dialCh       chan<- *Job
+	localCloseCh chan struct{}
+}
+
+var _ Throttler = (*throttler)(nil)
+
+func (t *throttler) Start(ctx context.Context, inCh <-chan *Job, dialCh chan<- *Job) {
+	t.inCh = inCh
+	t.dialCh = dialCh
+
+ThrottleLoop:
+	for {
+		// Process any token releases to avoid those channels getting backed up.
+		select {
+		case id := <-t.peerTokenFreed:
+			waitlist := t.waitingOnPeerLimit[id]
```

Comment:
it looks like we have a case where jobs waiting on the per-peer limit can then bypass the FD limits since we don't check here

Reply:
I don't this is possible because when we get a new job, we check the peer limit first, then the FD limit (by virtue of the implicit ordering of the `&&` [here](https://github.com/libp2p/go-libp2p-swarm/pull/88/files#diff-19c761b46b0f2b69d46a550b47c6ba6dR87)). If a job fails the peer limit, we queue until we get a slot. Then we check the FD limit before actually doing the dial.